src/com/yashu/interview/map/linkedhashset/HashSet.java

```diff
@@ -0,0 +1,218 @@
+package HashSet;
+import java.util.InputMismatchException;
+public class HashSet<T> implements OrderedSet<T>, Container<T>{
+
+	private	Node<T>[] array;
+	private ArrayIndexCalculator indexCalculator;
+	private LinkedList<T> linkedList;
+	private int count = 0;
+
+
+	/**
+	 * Parameterized constructor
+	 * @param initialSize - Initial size of the set
+	 */
+	@SuppressWarnings("unchecked")
+	public HashSet(int initialSize){
+		array = new Node[initialSize];
+		indexCalculator = new ArrayIndexCalculator();
+		linkedList = new LinkedList<T>();
+	}
+	 
+	// Package visible only - for testing purposes
+	@SuppressWarnings("unchecked")
+	HashSet(int initialSize, ArrayIndexCalculator indexCalculator) {
+		array = new Node[initialSize];
+		this.indexCalculator = indexCalculator;
+		linkedList = new LinkedList<T>();
+	}
+	// Returns the size of the Array.
+	public int getLengthOfArray() {
+		return array.length;
+	}
+
+	@SuppressWarnings("unchecked")
+	private void createNewArray(int length) {
+		array = new Node[length];
+	}
+	
+	//Iterator
+	private class LinkedHashSetIterator implements MyIterator<T>{
+		private Node<T> temp ;
+		
+		public LinkedHashSetIterator(){
+		temp = null;
+		}
+		@Override
+		public boolean hasNext() {
+			return temp!=linkedList.getTail();
+		}
+
+		@Override
+		public Object next() {
+			if(temp == null){
+				temp = linkedList.getHead() ;
+				return temp.getData();
+			}
+			temp = temp.getNext();
+			if(this.hasNext()){
+				return temp.getData();
+			}
+			return temp.getData();
+		}
+		
+	}
+	
+	@Override
+	public MyIterator<T> iterator() {
+		return new LinkedHashSetIterator();
+	}
+	
+	public void insert(T value) {
+		if(value == null){
+			throw new IllegalArgumentException();
+		}
+		//Checks uniqueness of the set.
+		if(exists(value)){return;}
+		//Doubles the size of Array if it is full.
+		if(count == array.length){
+			array = doubleArray(array.length);
+		}
+		count++;
+		Node<T> newNode = new Node<T>(value);
+		//Using the Hash Function the index of the node is determined.
+		int indexOfString = hashFunction(value);
+		//If the index is not null, collision is avoided using LinkedList. 
+		if(collision(value)){
+			linkedList.setTail(avoidCollision(value,newNode,linkedList.getTail(),indexOfString));
+			count--;
+			return;
+		}
+		insertIntoArray(indexOfString, newNode);
+		linkedList.insert(newNode);
+	}
+
+
+	Node<T>[] doubleArray(int length) {
+		@SuppressWarnings("unchecked")
+		Node<T>[] temp = new Node[length];
+		for(int i = 0;i<length;i++){
+			temp[i] = array[i];
+		}
+		createNewArray(2*length);
```

Comment:
You don't need to create a new temp array with existing size. You can simply do...

Node<T>[] temp = array;
createNewArray(2*length);